### PR TITLE
Issue/463 solver defs

### PR DESCRIPTION
**Pull request by pjljvandelaar**

### CODE REVIEW

**Comment by capitanbatata** [comment]
There are a couple of failing tests:

Any - Test 0:
```text
> TXSSERVER CMD >>  INIT data/integration-test/Any/anyTest.txs
> txsserver: Illegal input valexprToSMT - ValExpr {view = Vpredef SSS (FuncId {name = "dropWhile", unid = 535, funcargs = [SortId {name = "String", unid = 104},SortId {name = "String", unid = 104}], funcsort = SortId {name = "String", unid = 104}}) [ValExpr {view = Vvar (VarId {name = "x", unid = 558, varsort = SortId {name = "String", unid = 104}})},ValExpr {view = Vvar (VarId {name = "y", unid = 559, varsort = SortId {name = "String", unid = 104}})}]}
```

Process 12 - Unique ID #long:

```text
> TXSSERVER CMD >>  INIT examps/DispatchProcess/spec/DisPro12-unique-id.txs
> txsserver: Illegal input valexprToSMT - ValExpr {view = Vpredef SSS (FuncId {name = "dropWhile", unid = 535, funcargs = [SortId {name = "String", unid = 104},SortId {name = "String", unid = 104}], funcsort = SortId {name = "String", unid = 104}}) [ValExpr {view = Vvar (VarId {name = "x", unid = 558, varsort = SortId {name = "String", unid = 104}})},ValExpr {view = Vvar (VarId {name = "y", unid = 559, varsort = SortId {name = "String", unid = 104}})}]}
```

Multiple Loops Stepper #long

```text
> txsserver: Illegal input valexprToSMT - ValExpr {view = Vpredef SSS (FuncId {name = "dropWhile", unid = 535, funcargs = [SortId {name = "String", unid = 104},SortId {name = "String", unid = 104}], funcsort = SortId {name = "String", unid = 104}}) [ValExpr {view = Vvar (VarId {name = "x", unid = 558, varsort = SortId {name = "String", unid = 104}})},ValExpr {view = Vvar (VarId {name = "y", unid = 559, varsort = SortId {name = "String", unid = 104}})}]}
```

File: sys/solve/src/RandPartition.hs [hunk 1]
```diff
@@ -218,7 +218,14 @@ randCnrs p vexp depth  =
                                 else if  depth > 0
                                        then randCnrsADT p vexp depth
                                        else return [ Set.empty ]
- 
+
+-- ----------------------------------------------------------------------------------------- --
+-- cartesian product
+
+cartProd :: [[t]] -> [[t]]
```
Comment (capitanbatata) [inline sys/solve/src/RandPartition.hs]: Did we define tests for it?

File: sys/valexpr/src/ConstDefs.hs [hunk 2]
```diff
@@ -31,3 +32,12 @@ data Const = Cbool    { cBool :: Bool }
   deriving (Eq, Ord, Read, Show, Generic, NFData, Data)
 
 instance Resettable Const
+
+instance SortOf Const where
+  sortOf (Cbool _b)                        = sortId_Bool
```
Comment (capitanbatata) [inline sys/valexpr/src/ConstDefs.hs]: Why not using plain wild cards?

File: sys/solve/src/RandPartition.hs (hunk 1 above)
Comment (pjljvandelaar) [reply sys/solve/src/RandPartition.hs]: moved from defs.Utils

File: sys/valexpr/src/ConstDefs.hs (hunk 2 above)
Comment (pjljvandelaar) [reply sys/valexpr/src/ConstDefs.hs]: just a copy...
Probably 
1. was b in the past.
 2. hlint complaints b isnt used
3. add _ to silence hlint

Comment (pjljvandelaar) [reply sys/valexpr/src/ConstDefs.hs]: When one has multiple unused arguments, and want to use one names help in picking the right one.

**Comment by pjljvandelaar** [comment]
hlint: no new issues, only moved definitions!

**Comment by pjljvandelaar** [comment]
Due to the renaming of SortId_<Type> to SortId<Type>, code climate sees both the removal and addition of issues....
Yet the number of issues is going down...

**Comment by pjljvandelaar** [comment]
Had  to go to the fysio....
Must be map.difference i.s.o. map.diff
Will change it tomorrow

**Comment by pjljvandelaar** [comment]
All checks have passed ;-)
Nicely before the end of the year .....

**Comment by pjljvandelaar** [comment]
@capitanbatata I expect no impact on the performance... Can you check this?

**Comment by keremispirli** [comment]
I ran the benchmarks.

`Hiding/match Int` is slowed down by a factor of **17.66**.
Performance for some of the others improved up to 14%, while some others worsened up to 10%. 

[Details](https://docs.google.com/spreadsheets/d/1vmFDb21beeJFvwyN_GpSXbu3y3gMfcOb5F77omONqDk/edit#gid=638985914)
[Comparison](https://docs.google.com/spreadsheets/d/1vmFDb21beeJFvwyN_GpSXbu3y3gMfcOb5F77omONqDk/edit#gid=142689038)
[Plots](https://docs.google.com/spreadsheets/d/1vmFDb21beeJFvwyN_GpSXbu3y3gMfcOb5F77omONqDk/edit#gid=139218297)

Output of `Hiding/match Int`:
```
benchmarking Hiding/match Int
................time                 75.13 s    (72.58 s .. 76.57 s)
                     1.000 R²   (1.000 R² .. NaN R²)
mean                 75.31 s    (74.91 s .. 75.52 s)
std dev              343.3 ms   (0.0 s .. 363.0 ms)
variance introduced by outliers: 19% (moderately inflated)
```

I ran the benchmarks twice and second time was during night, which means this result is not a hiccup.

**Comment by pjljvandelaar** [comment]
totally unexpected....
Is it limited to your pc only?

**Comment by keremispirli** [comment]
@pjljvandelaar We can run another benchmark when @capitanbatata's machine arrives, to see if that's the case.

**Pull request merged**